src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1041,102 +1041,184 @@ private bool IsValidFixedVariableInitializer(TypeSymbol declType, SourceLocalSym
-                }
-                else if (initializerType.IsArray())
-                {
-                    // See ExpressionBinder::BindPtrToArray (though most of that functionality is now in LocalRewriter).
-                    var arrayType = (ArrayTypeSymbol)initializerType;
-                    elementType = arrayType.ElementType;
+                case BoundKind.FieldAccess:
+                    var fa = (BoundFieldAccess)initializerOpt;
+                    if (fa.FieldSymbol.IsFixed)
+                    {
+                        elementType = ((PointerTypeSymbol)fa.Type).PointedAtType;
+                        break;
+                    }
+
+                    goto default;
+
+                default:
+                    //  fixed (T* variable = <expr>) ...
 
-                    if (elementType.IsManagedType)
+                    // check for arrays
+                    if (initializerType.IsArray())
                     {
-                        Error(diagnostics, ErrorCode.ERR_ManagedAddr, initializerSyntax, elementType);
-                        hasErrors = true;
+                        // See ExpressionBinder::BindPtrToArray (though most of that functionality is now in LocalRewriter).
+                        var arrayType = (ArrayTypeSymbol)initializerType;
+                        elementType = arrayType.ElementType;
+                        break;
                     }
-                }
-                else if (!initializerOpt.HasAnyErrors)
-                {
-                    elementType = initializerOpt.Type;
-                    switch (initializerOpt.Kind)
+
+                    // check for a special ref-returning method
+                    var additionalDiagnostics = DiagnosticBag.GetInstance();
+                    try
                     {
-                        case BoundKind.AddressOfOperator:
-                            elementType = ((BoundAddressOfOperator)initializerOpt).Operand.Type;
+                        getPinnableMethod = GetDangerousGetPinnableReferenceMethod(initializerOpt, additionalDiagnostics);
+
+                        bool extensisbleFixedEnabled = ((CSharpParseOptions)initializerOpt.SyntaxTree.Options)?.IsFeatureEnabled(MessageID.IDS_FeatureExtensibleFixedStatement) != false;
+
+                        // check for String
+                        // NOTE: We will allow DangerousGetPinnableReferenceMethod to take precendence, but only if it is a member of System.String
+                        if (initializerType.SpecialType == SpecialType.System_String &&
+                            ((object)getPinnableMethod == null || !extensisbleFixedEnabled || getPinnableMethod.ContainingType.SpecialType != SpecialType.System_String))
+                        {
+                            elementType = this.GetSpecialType(SpecialType.System_Char, diagnostics, initializerSyntax);
+                            Debug.Assert(!elementType.IsManagedType);
```

Comment:
I don't understand what this assert is supposed to protect against. If the user code is wrong, why would we assert instead of provide an error?

Reply:
useless assert. it will always pass. removed.

---
In reply to: [167708843](https://github.com/dotnet/roslyn/pull/24469#discussion_r167708843) [](ancestors = 167708843)